README.md

```diff
@@ -1,2 +1,175 @@
+| QmOne | <------------------------------------> | QmRelay | <---------------------------> | QmTwo |
++-------+                                        +---------+                               +-------+
+      ^                                            +-----+                                    ^
+      |                                            |     |                                    |
+      |        /libp2p-circuit-relay/QmTwo         |     |                                    |
+      +--------------------------------------------+     +------------------------------------+
+```
+
+[go-libp2p-nat]: https://github.com/libp2p/go-libp2p-nat
+
+
+## Dramatization
+
+Cast:
+- QmOne, the dialing node (browser).
+- QmTwo, the listening node (go-ipfs).
+- QmRelay, a node which speaks the circuit relay protocol (go-ipfs or js-ipfs).
+
+Scene 1:
+- QmOne wants to connect to QmTwo,
+  and through peer routing has acquired a set of addresses of QmTwo.
+- QmTwo doesn't support any of the transports used by QmOne.
+- Awkward silence.
+
+Scene 2:
+- All three nodes have learned to speak the `/ipfs/relay/circuit` protocol.
+- QmRelay is configured to allow relaying connections between other nodes.
+- QmOne is configured to use QmRelay for relaying.
+- QmOne automatically added `/libp2p-circuit-relay/ipfs/QmTwo` to its set of QmTwo addresses.
+- QmOne tries to connect via relaying, because it shares this transport with QmTwo.
+- A lively and prolonged dialogue ensues.
+
+
+### Addressing
+
+Q: should /relay encapsulate /ipfs?
+  - such that /libp2p-circuit-relay/ipfs/Qmfoobar
+  - might be mandated by relay-as-a-transport...
+  - A: yes
+  - need to update the multiaddr size from V to 0
+Q: should the relay transport be configured a a simple listener?
+  - e.g. simply on /libp2p-circuit-relay
+  - or should it just be enabled by default?
+
+Relay addresses can be encapsulated "infinitely",
+so that one can construct onion-like relay chains:
+`/libp2p-circuit-relay/ipfs/QmFoo/libp2p-circuit-relay/ipfs/QmBar`
+
+Relay addresses can encapsulated in an /ipfs address to signal a specific relay node to use:
+`/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
+or `/dns6/relay.libp2p.io/ipfs/QmRelay/libp2p-circuit-relay/ipfs/QmFoo`
+
+
+### /ipfs/relay/circuit stream protocol
+
+Each relayed connection corresponds to two multistreams,
+one between QmOne and QmRelay, the other between QmRelay and QmTwo.
+
+QmOne writes:
+
+```
+/ipfs/relay/circuit/0.1.0
+QmOne
+QmTwo
+inner connection data from here on, most likely multistream-select, then secio
+```
+
+```
+<multistream-header>
+<src>
+<dst>
+<data>
+```
+
+- `multistream-header`: This field is not strictly part of the relay stream.
+  - Multistreams are prefixed with a multicodec field identifying the protocol used for the stream.
+- `src`: The PeerID of the node opening the connection.
+- `dst`: The PeerID of the node that `src` wants to connect to.
```

Comment:
hmmm i may be fine if this is a peer-only thing. then it works differently than the circuit relay thing i had made, which dials to multiaddrs. (i.e. i tell A what to dial to connect to the next step)